flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -83,24 +86,45 @@
 	/** The total number of bytes (both data and event buffers). */
 	private long totalNumberOfBytes;
 
+	/** The collection of buffers which are spanned over by checkpoint barrier and needs to be persisted for snapshot. */
+	private final List<Buffer> inflightBufferSnapshot = new ArrayList<>();
+
 	// ------------------------------------------------------------------------
 
 	PipelinedSubpartition(int index, ResultPartition parent) {
 		super(index, parent);
 	}
 
 	@Override
-	public boolean add(BufferConsumer bufferConsumer) {
-		return add(bufferConsumer, false);
+	public boolean add(BufferConsumer bufferConsumer) throws IOException {
+		return add(bufferConsumer, false, false);
 	}
 
 	@Override
 	public void finish() throws IOException {
-		add(EventSerializer.toBufferConsumer(EndOfPartitionEvent.INSTANCE), true);
+		add(EventSerializer.toBufferConsumer(EndOfPartitionEvent.INSTANCE), true, false);
 		LOG.debug("{}: Finished {}.", parent.getOwningTaskName(), this);
 	}
 
-	private boolean add(BufferConsumer bufferConsumer, boolean finish) {
+	@Override
+	public EnqueueAction priorityEvent(AbstractEvent event, BufferConsumer eventBufferConsumer) throws IOException {
+		if (readView != null) {
+			switch (readView.priorityEvent(event, eventBufferConsumer)) {
+				case ENQUEUE_AS_FIRST_BUFFER:
+					add(eventBufferConsumer.copy(), false, true);
+					break;
+				case ENQUEUE_AS_LAST_BUFFER:
+					add(eventBufferConsumer.copy(), false, false);
+					break;
+			}
+		} else {
+			buffers.add(eventBufferConsumer.copy());
```

Comment:
If the current view is null, maybe it is also worth inserting the barrier into head, because the view might be established soon after inserting. Or are there any considerations for making the difference based on the view status?